syntax = "proto3";

package photonvision.proto;

option java_package = "org.photonvision.proto";

import "geometry3d.proto";

message ProtobufTargetCorner {
  double x = 1;
  double y = 2;
}

message ProtobufPNPResult {
  bool isPresent = 1;
  wpi.proto.ProtobufTransform3d best = 2;
  double bestReprojErr = 3;
  optional wpi.proto.ProtobufTransform3d alt = 4;
  optional double altReprojErr = 5;
  double ambiguity = 6;
}

message ProtobufMultiTargetPNPResult {
  ProtobufPNPResult estimatedPose = 1;
  repeated int32 fiducial_ids_used = 2;
}

message ProtobufPhotonTrackedTarget {
  double yaw = 1;
  double pitch = 2;
  double area = 3;
  double skew = 4;
  int32 fiducialId = 5;
  wpi.proto.ProtobufTransform3d bestCameraToTarget = 6;
  wpi.proto.ProtobufTransform3d altCameraToTarget = 7;
  double poseAmbiguity = 8;
  repeated ProtobufTargetCorner minAreaRectCorners = 9;
  repeated ProtobufTargetCorner detectedCorners = 10;
}

message ProtobufPhotonPipelineResult {
  double latency_ms = 1;

  repeated ProtobufPhotonTrackedTarget targets = 2;
  ProtobufMultiTargetPNPResult multi_target_result = 3;
}
